libs/langchain/langchain/automaton/chat_agent.py

```diff
@@ -0,0 +1,126 @@
+
+Provide only ONE action per $BLOB, as shown.
+
+<action>
+{{
+  "action": $TOOL_NAME,
+  "action_input": $INPUT
+}}
+</action>
+
+When invoking a tool do not provide any clarifying information.
+
+The human will forward results of tool invocations as "Observations".
+
+When you know the answer paraphrase the information in the observations properly and respond to the user. \
+If you do not know the answer use more tools.
+
+You can only take a single action at a time."""
+)
+
+
+def generate_prompt(tools: Sequence[BaseTool]) -> MessageLog:
+    """Generate a prompt for the agent."""
+    tool_info = generate_tool_info(tools)
+    msg = TEMPLATE_.format(**tool_info)
+    return MessageLog(messages=[msg])
+
+
+def decode(text: Union[BaseMessage, str]) -> MessageLike:
+    """Decode the action."""
+    pattern = re.compile(r"<action>(?P<action_blob>.*?)<\/action>", re.DOTALL)
+    if not isinstance(text, BaseMessage):
+        raise NotImplementedError()
+    _text = text.content
+    match = pattern.search(_text)
+    if match:
+        action_blob = match.group("action_blob")
+        data = ast.literal_eval(action_blob)
+        name = data["action"]
+        if name == "Final Answer":  # Special cased "tool" for final answer
+            return AgentFinish(result=data["action_input"])
+        return FunctionCall(name=data["action"], arguments=data["action_input"] or {})
+    else:
+        return AgentFinish(result=text)
+
+
+def prompt_generator(log: MessageLog) -> List[BaseMessage]:
+    """Generate a prompt from a log of message like objects."""
+    messages = []
+    for message in log.messages:
+        if isinstance(message, BaseMessage):
+            messages.append(message)
+        elif isinstance(message, FunctionResult):
+            messages.append(
+                HumanMessage(
+                    content=f"Observation: {message.result}",
+                )
+            )
+        else:
+            pass
+    return messages
+
+
+class ChatAgent:
+    """An agent for chat models."""
+
+    def __init__(
+        self,
+        llm: BaseLanguageModel,
+        tools: Sequence[BaseTool],
+        *,
+        max_iterations: int = 10,
+    ) -> None:
+        """Initialize the chat automaton."""
+        self.llm_program = create_llm_program(
+            llm,
+            prompt_generator=prompt_generator,
+            tools=tools,
+            parser=decode,
+        )
+        self.max_iterations = max_iterations
+
+    def run(self, message_log: MessageLog) -> None:
+        """Run the agent."""
+        if not message_log:
+            raise AssertionError(f"Expected at least one message in message_log")
+
+        for _ in range(self.max_iterations):
+            last_message = message_log[-1]
+
+            if isinstance(last_message, AgentFinish):
+                break
+
+            messages = self.llm_program.invoke(message_log)
+            message_log.add_messages(messages)
```

Comment:
What's the advantage of using callbacks instead of an Iterator? I understand you want to accumulate messages somewhere, but worth yielding them too?